File: test-security.mjs
import assert from 'node:assert'

const BASE = 'http://localhost:3000/api/secure-endpoint'
const PASS = '\x1b[32m✓ PASS\x1b[0m'
const FAIL = '\x1b[31m✗ FAIL\x1b[0m'

// ─── Test A: The Zod Shield ─────────────────────────────────────────────────
async function testZodShield() {
  console.log('\n── Test A: Zod Validation Shield ──')

  const malformedPayload = {
    email: 'not-an-email',   // bad format
    // message is missing
    count: 'not-a-number',   // wrong type
    extraField: 'should be stripped',
  }

  const res = await fetch(BASE, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(malformedPayload),
  })

  const json = await res.json()

  assert.strictEqual(res.status, 400, `Expected 400, got ${res.status}`)
  assert.ok(json.details, 'Expected validation error details')
  assert.ok(json.details.email, 'Expected email field error')
  assert.ok(json.details.message, 'Expected message field error')
  assert.ok(json.details.count, 'Expected count field error')

  console.log(`${PASS}  Malformed payload rejected with 400 + field errors`)
}

// ─── Test B: The API Leak Test ──────────────────────────────────────────────
async function testApiLeak() {
  console.log('\n── Test B: Secret Leak Prevention ──')

  const validPayload = {
    email: 'test@example.com',
    message: 'Hello from the stress test',
    count: 42,
  }

  const res = await fetch(BASE, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(validPayload),
  })

  const json = await res.json()
  const bodyStr = JSON.stringify(json)

  assert.strictEqual(res.status, 200, `Expected 200, got ${res.status}`)
  assert.ok(!bodyStr.includes('INTERNAL_API_SECRET'), 'Response must not contain secret name')
  assert.ok(!bodyStr.includes('supersecretkey'), 'Response must not contain secret value')

  console.log(`${PASS}  200 OK returned — no secrets leaked to client`)
}

// ─── Test C: DDoS Simulation ────────────────────────────────────────────────
async function testRateLimit() {
  console.log('\n── Test C: Rate-Limit DDoS Simulation ──')

  const validPayload = { email: 'ddos@test.com', message: 'ratelimit test', count: 1 }
  const results = []

  // Fire 8 requests in parallel
  const promises = Array.from({ length: 8 }, () =>
    fetch(BASE, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(validPayload),
    }).then((r) => r.status)
  )

  const statuses = await Promise.all(promises)
  statuses.forEach((s, i) => results.push({ req: i + 1, status: s }))

  const allowed = results.filter((r) => r.status === 200)
  const blocked = results.filter((r) => r.status === 429)

  assert.ok(allowed.length <= 5, `Expected ≤5 allowed, got ${allowed.length}`)
  assert.ok(blocked.length >= 3, `Expected ≥3 blocked, got ${blocked.length}`)

  results.forEach((r) => {
    const icon = r.status === 200 ? PASS : r.status === 429 ? PASS : FAIL
    console.log(`  ${icon}  Request ${r.req} → ${r.status}`)
  })

  console.log(`${PASS}  ${allowed.length} allowed, ${blocked.length} blocked (429)`)
}

// ─── Runner ─────────────────────────────────────────────────────────────────
;(async () => {
  console.log('╔══════════════════════════════════════════╗')
  console.log('║   OWASP Security Layer Stress Test       ║')
  console.log('╚══════════════════════════════════════════╝')

  try {
    await testZodShield()
    await testApiLeak()
    await testRateLimit()

    console.log('\n\x1b[36mAll tests passed.\x1b[0m')
  } catch (err) {
    console.error(`\n${FAIL}  ${err.message}`)
    process.exit(1)
  }
})()

File: backend/test_security.py
import asyncio
import json
import sys
import httpx

# Test through the proxy (the real secure path)
BASE = "http://localhost:3001"
PASS = "[PASS]"
FAIL = "[FAIL]"


async def test_zod_shield():
    print("\n-- Test A: Pydantic Validation Shield --")

    malformed = {
        "email": "not-an-email",
        "count": "not-a-number",
        "extraField": "should be dropped",
    }

    async with httpx.AsyncClient() as client:
        res = await client.post(f"{BASE}/api/leads", json=malformed)
        body = res.json()

    # Proxy converts FastAPI's 422 into a 400 with sanitized error payload
    assert res.status_code == 400, f"Expected 400, got {res.status_code}"
    assert body.get("status") == "validation_error", "Expected validation_error status"
    assert body.get("details") is not None, "Expected validation error details"
    print(f"{PASS}  Malformed payload rejected with 400 + field errors")
    print(f"       Errors: {json.dumps(body.get('details', body), indent=2)}")


async def test_api_leak():
    print("\n-- Test B: Secret Leak Prevention --")

    valid = {
        "name": "Test User",
        "email": "test@example.com",
        "message": "Hello from the stress test",
        "project_tier": "The Velocity Build",
    }

    async with httpx.AsyncClient() as client:
        res = await client.post(f"{BASE}/api/leads", json=valid)
        body = res.json()
        body_str = json.dumps(body)

    # Proxy returns status=201 with { status: "success", data: { lead fields } }
    assert res.status_code == 201, f"Expected 201, got {res.status_code}"
    assert body.get("status") == "success", "Expected success status"
    assert body.get("data") is not None, "Expected data payload"
    data = body["data"]
    assert "INTERNAL_API_SECRET" not in body_str, "Response must not contain secret name"
    assert "supersecretkey" not in body_str, "Response must not contain secret value"
    assert "id" in data, "Response must contain lead id"
    assert "name" in data, "Response must contain lead name"
    print(f"{PASS}  201 Created returned -- no secrets leaked to client")


async def test_rate_limit():
    print("\n-- Test C: Rate-Limit DDoS Simulation --")

    valid = {
        "name": "DDoS Test",
        "email": "ddos@test.com",
        "message": "ratelimit test",
    }

    async with httpx.AsyncClient() as client:
        tasks = [client.post(f"{BASE}/api/leads", json=valid) for _ in range(8)]
        responses = await asyncio.gather(*tasks)

    statuses = [r.status_code for r in responses]
    allowed = statuses.count(201)
    blocked = statuses.count(429)

    for i, s in enumerate(statuses, 1):
        icon = PASS if s in (201, 429) else FAIL
        print(f"  {icon}  Request {i} -> {s}")

    assert allowed <= 5, f"Expected <=5 allowed, got {allowed}"
    assert blocked >= 3, f"Expected >=3 blocked, got {blocked}"
    print(f"{PASS}  {allowed} allowed, {blocked} blocked (429)")


async def main():
    print("=" * 48)
    print("  OWASP Security Layer Stress Test (FastAPI)")
    print("=" * 48)

    try:
        await test_zod_shield()
        await test_api_leak()
        await test_rate_limit()
        print("\nAll tests passed.")
    except AssertionError as e:
        print(f"\n{FAIL}  {e}")
        sys.exit(1)
    except httpx.ConnectError:
        print(f"\n{FAIL}  Cannot connect to {BASE} -- is the server running?")
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())
